### ISSUE TITLE
Zigbee2mqtt to fast for adapt 

### ISSUE THREAD

**Issue by R3Cardoso** [basnijholt/adaptive-lighting#996]
Hi, this is not a bug, but I wasn't sure where to put my question.

I have Philips Hue devices on my HA working prefect with AL.
But I decided to change my devices to zigbee2mqtt and noticed a improve on response time, for exemple when I click on a light switch and the light turns on.
But this come with a problem. AL seems to not be able to adapt lights before they turn on.
With Philips Hue intregation it works fine, but it take a 1 second to respond to the click.

Is there any setting on AL or zigbee2mqtt that resolve this?

**Comment by th3w1zard1** [comment]
https://github.com/basnijholt/adaptive-lighting/discussions/501

**Comment by Kuchiru** [comment]
I think i have the same issue so i'll add the required info as i'm interested if there is a solution to this as well.

For me the best example of this is this:
- I have some motion activated hue lights, these change color temperature as set by AL.
- Sleep mode is automatically activated at 22:30
- If the motion activated light turns off before 22:30 and is activated after 22:30 in the evening or in the night it will blind you before adjusting to the sleep mode brightness which usually takes about 1 second.

The lamps used for the example are https://www.zigbee2mqtt.io/devices/9290022169.html

z2m configuration.yaml
`homeassistant:
  legacy_entity_attributes: true
advanced:
  network_key:
    - 98
    - 146
    - 27
    - 216
    - 99
    - 152
    - 158
    - 208
    - 145
    - 36
    - 33
    - 19
    - 122
    - 31
    - 53
    - 223
  pan_id: 30227
  ext_pan_id:
    - 245
    - 248
    - 22
    - 74
    - 178
    - 205
    - 141
    - 150
  homeassistant_legacy_entity_attributes: false
  legacy_api: false
  legacy_availability_payload: false
  cache_state: true
  cache_state_persistent: true
  cache_state_send_on_startup: false
  report: true
mqtt:
  server: mqtt://core-mosquitto:1883
  user: addons
  password: -
  force_disable_retain: false
frontend:
  port: 8099
serial:
  port: >-
    /dev/serial/by-id/usb-Silicon_Labs_slae.sh_cc2652rb_stick_-_slaesh_s_iot_stuff_00_12_4B_00_23_90_B8_CA-if00-port0
device_options:
  legacy: false
  disabled: false
  homeassistant: true
  optimistic: true
  filtered_cache:
    - brightness
    - rgb_color
    - hs_color
    - xy_color
    - color_temp_kelvin
    - color_temp
    - color
    - color_mode
devices:
  '0x0017880102b0a892':
    friendly_name: ambiance_lamp
    transition: 1
  '0x001788010b7ed23f':
    friendly_name: couch_ceiling_light
    transition: 1
  '0x001788010b05f506':
    friendly_name: dining_table_ceiling_light
    transition: 1
  '0x001788010b553f41':
    friendly_name: kitchen_ceiling_light_1A
    transition: 1
  '0x001788010b501c7c':
    friendly_name: kitchen_ceiling_light_1B
    transition: 1
  '0x001788010901438e':
    friendly_name: bedroom_ceiling_light
    transition: 1
  '0x00178801062eeb48':
    friendly_name: passage_ceiling_light_1A
    transition: 1
  '0x0017880103282dea':
    friendly_name: hallway_motion_sensor
  '0x00178801062f06b9':
    friendly_name: passage_ceiling_light_1B
    transition: 1
  '0x0017880104876af2':
    friendly_name: Colour bulb 1
    transition: 1
  '0x00178801061d77b1':
    friendly_name: tv_light
    transition: 1
  '0x0017880106a07530':
    friendly_name: Colour bulb 3
    transition: 1
  '0x54ef441000a8c602':
    friendly_name: kitchen_motion_sensor
    occupancy_timeout: 60
  '0x0017880108c232a9':
    friendly_name: backyard_lamp
  '0x0017880108c23366':
    friendly_name: frontdoor_lamp
  '0xa4c138026f9adda5':
    friendly_name: kitchen_countertop_light
    transition: 1
  '0x001788010df5a852':
    friendly_name: passage_motion_sensor
  '0x187a3efffe6dabbb':
    friendly_name: '0x187a3efffe6dabbb'
    humidity_calibration: 0
    temperature_calibration: 0
    pm25_calibration: 0
  '0x001788010d7b6acc':
    friendly_name: hallway_ceiling_lamp
    transition: 1
  '0x54ef441000a12b7d':
    friendly_name: Supply Closet Door Sensor
  '0x8cf681fffe504b8d':
    friendly_name: Mailbox Sensor
    occupancy_timeout: 60
  '0x0017880104f3f003':
    friendly_name: '0x0017880104f3f003'
  '0x00178801062eecaf':
    friendly_name: office_ceiling_light
groups:
  '1':
    friendly_name: kitchen_ceiling_light
    devices:
      - 0x001788010b553f41/11
      - 0x001788010b501c7c/11
  '2':
    friendly_name: passage_ceiling_light
    devices:
      - 0x00178801062eeb48/11
      - 0x00178801062f06b9/11
`

**Comment by djurny** [comment]
Hi there,
I cannot make out if you are talking about AL not adapting the color or the brightness of your lights.
If it's color adaptation, do try if the combination of `execute_if_off` and setting the `color_temp` before turning on the lights will solve your issue (https://github.com/basnijholt/adaptive-lighting/issues/1022#issuecomment-2218830648).
Also check if you might have set `adapt_delay` in your AL configuration, if you want AL to adapt as quickly as possible, make sure that `switch.adaptive_lighting_<name>`, `switch.adaptive_lighting_adapt_brightness_<name>` and `switch.adaptive_lighting_adapt_color_<name>` are on, `manual_control` is off and the `adapt_delay` is set to '0'.
Fiddling with `intercept` and `multi_light_intercept` might also work for you, as that tries to put all the settings you want in the `light.turn_on` (also take `only_once` and `adapt_only_on_bare_turn_on` into account) call from HASS.
Groetjes,

**Comment by djurny** [comment]
Hi there,
Copied the reply from https://github.com/basnijholt/adaptive-lighting/discussions/506#top here, as it seems your question originated from this issue thread.

If your issue is that the color temperature of your lights is not changing immediately after turning on your bulb, this reply is for you.

In order to understand your issue better, the following information will help:
- what method of switching off and switching on the light are you using?
e.g. using 'empty' `light.turn_on` or a `light.turn_on` with `brightness`, `color_temp` or other option.
- what do you expect to see and what is your observation? 
e.g. brightness does not change/transition after power on? or `color`/`color_temp` not change/transition? or `brightness`/`color` does change, but only after a few seconds or after a full `interval` as specified in the `adaptive_lighting` entity?

Any detail helps.

> Thanks for the help, but I wasn't able to resolve anything. I'm not sure where I need to put the code? should be on all automations? `//snip - repeat: for_each: '{{ state_attr(entity_light, "entity_id") }}' sequence: - service: mqtt.publish data_template: topic: 'zigbee2mqtt/{{state_attr(repeat.item, "friendly_name")}}/set/color_options' payload: | { "execute_if_off": "true" } - service: mqtt.publish data_template: topic: 'zigbee2mqtt/{{state_attr(repeat.item, "friendly_name")}}/set' payload: | { "color_temp": "{{turn_on_quirk_color_temp_mired_current}}", "transition": "0" } //snip`
> 
> About transition I got 0 on zigbee2mqtt and adapt_delay, on initial_transition got 1 because it helps disguise the change.

To 'prime' the `color_temp` of your lights, which means set them to the wanted `color_temp` _before_ you turn them on, you would have to make a script in homeassistant. You can read up [here](https://www.home-assistant.io/integrations/script/) on what that entails.

As an basic example, your script would look as follows (this is a dry example, have not tested it whatsoever):

```
script:
  some_script_name_that_makes_sense:
    variables:
      ## change these to suit your situation
      name_light: 'kitchen'
      name_adaptive_lighting: 'kitchen'
      transition: '1'
      ## 
      entity_id_light: 'light.{{name_adaptive_lighting}}'
      entity_id_adaptive_lighting: 'adaptive_lighting.{{name_adaptive_lighting}}'
      mqtt_topic: 'zigbee2mqtt/{{name_light}}/set'
    sequence:
    - action: mqtt.publish
      data_template:
        payload: |-
            {
              "color_temp": "{{state_attr(entity_id_adaptive_lighting, 'color_temp_mired')}}"
            }
        topic: '{{mqtt_topic}}'
    - delay:
        hours: 0
        minutes: 0
        seconds: 0
        milliseconds: 250
    - light.turn_on
      data:
        transition: '{{transition}}'
      target:
        entity_id: '{{entity_id_light}}'
```

Note that for the script to work:
- Your light should be a device/entity and not a group in home assistant. To make the script work for light groups, some additional YAML is required, but I wanted to make the example as short and easy as it can be.
- Your light's name in home assistant should be the same as the `friendly_name` in zigbee2mqtt.
- The adaptive lighting entity must have the same name as your light's name.
- The switch entity `switch.adaptive_lighting_kitchen` switch must be enabled. (Normally speaking this is enabled unless you deliberately have turned it off.)
- The switch entity `switch.adaptive_lighting_adapt_color_kitchen` must be enabled. (Normally speaking this is enabled unless you deliberately have turned it off.)
- The switch entity `switch.adaptive_lighting_adapt_brightness_kitchen` must be enabled. (Normally speaking this is enabled unless you deliberately have turned it off.)

There is no simpler way to do this a.f.a.i.k.

Hope that helps,
Groetjes,

**Comment by Kuchiru** [comment]
Hi @djurny,

Thank you for the example and the in depth post.
I wonder if AL could actually update the color temperature for lights in the background even if the state is off, for lights that have the color options enabled.

I ended up switching to ZHA as this integration has execute_if_off implemented, but regret soon came.
For some reason it stopped working, or it is inconsistent which made me switch back to Z2M as i got a new coordinator - i needed to re-pair everything anyway.

I've gotten it this to work with the script/Z2M with some minor adjustments.
For anyone that finds this, here is my version:
```
alias: Set Light Color Temperature
fields:
  light_entity:
    name: Light
    selector:
      entity:
        domain: light
    description: Entity to control
    required: true
  adaptive_lighting_entity:
    name: Adaptive Lighting
    selector:
      entity:
        domain: switch
    description: Entity for the adaptive lighting configuration
    required: true
sequence:
  - variables:
      mqtt_topic: zigbee2mqtt/{{ light_entity.split('.')[-1] }}/set
  - action: mqtt.publish
    data_template:
      payload: |-
        {
          "color_temp": "{{ state_attr(adaptive_lighting_entity, 'color_temp_mired') }}",
          "transition": 0
        }
      topic: "{{ mqtt_topic }}"
  - delay:
      milliseconds: 250
    enabled: false
  - action: light.turn_on
    target:
      entity_id: "{{ light_entity }}"
description: ""
```

**Comment by djurny** [comment]
Hi @R3Cardoso,
Continuing here from [here](https://github.com/basnijholt/adaptive-lighting/discussions/506#discussioncomment-10760968).

Can you list down the following:
- Version of home assistant
- Version of the [zigbee2mqtt](https://github.com/zigbee2mqtt/hassio-zigbee2mqtt) plugin 
- Version of zigbee2mqtt
- Version of Mosquitto broker (mqtt)
- Your zigbee dongle
- Your zigbee dongle's firmware version
- The type of Ikea lights you have
- The firmware version of the Ikea lights you have

More to ask after this...
Grt,

**Comment by djurny** [comment]
Hi @Kuchiru,
I also made a new automation for home assistant that will walk through all adaptive lighting instances that are enabled and configure/set the `color_temp` when the lights are off: [here](https://gist.github.com/djurny/f3e8dfc429f48d312ac172f5ac5a2c01).
Grt,

**Comment by R3Cardoso** [comment]
**- Version of home assistant**
HA core  2025.3.3
HA OS 15.0
**- Version of the [zigbee2mqtt](https://github.com/zigbee2mqtt/hassio-zigbee2mqtt) plugin**
Zigbee2mqtt 2.1.3-1
**- Version of zigbee2mqtt**
Zigbee2mqtt 2.1.3-1
**- Version of Mosquitto broker (mqtt)**
Mosquitto broker 6.5.0
**- Your zigbee dongle**
Now SLZB-06Mg24, but with SLZB-06p7 or Skyconnect was the same.
**- Your zigbee dongle's firmware version**
Core v2.8.2.dev3 and Zigbee Coordinator 20250212
**- The type of Ikea lights you have**
None, it't all Philips Hue Ambiance Light models.
**- The firmware version of the Ikea lights you have**
[1.122.2](https://www.philips-hue.com/en-us/support/release-notes/lamps)

**Comment by djurny** [comment]
Hi @R3Cardoso,
> - The type of Ikea lights you have
> None, it't all Philips Hue Ambiance Light models.

Ah my bad. Can you note the actual version/description of your Hue bulbs? The ones I have here are:

- Zigbee Model LCD007 Hue White and Color Ambiance slim downlight 6"
  Model [579573](https://www.zigbee2mqtt.io/devices/579573.html#philips-579573)
  Firmware build date 20220502 version 1.93.11

- Zigbee Model LTA010 Hue white ambiance A19 1100lm with Bluetooth
  Model [9290024683](https://www.zigbee2mqtt.io/devices/9290024683.html#philips-9290024683)
  Firmware build date 20240902 version 1.122.2

- Zigbee Model LTA012 Hue white ambiance E26 1600lm with Bluetooth
  Model [929002335105](https://www.zigbee2mqtt.io/devices/929002335105.html#philips-929002335105)
  Firmware build date 20240902 version 1.122.2

Grt,

**Comment by R3Cardoso** [comment]
I have this models:
[9290024684](https://www.zigbee2mqtt.io/devices/9290024684.html#philips-9290024684)
[929001953301](https://www.zigbee2mqtt.io/devices/929001953301.html#philips-929001953301)
[8719514301542](https://www.zigbee2mqtt.io/devices/8719514301542.html#philips-8719514301542)
[8719514328242](https://www.zigbee2mqtt.io/devices/8719514328242.html#philips-8719514328242)
[8719514392830](https://www.zigbee2mqtt.io/devices/8719514392830.html#philips-8719514392830)

They are all on same firmware version I mention before.

**Comment by djurny** [comment]
Hi @R3Cardoso,
Let's focus on (one of your) [9290024684](https://www.zigbee2mqtt.io/devices/9290024684.html#philips-9290024684) bulbs first. Judging by the model codes, it's perhaps a 220V version of the 110V version I have?

1. Make sure nothing is talking to the bulb.
Are you using the Hue app on any device? If so, either disable Wifi or Bluetooth on that device to make sure only zigbee2mqtt is talking to the bulb.
Are you using the Hue integration in home assistant? If so, disable it for now.

1. Delete the bulb from zigbee2mqtt.
Use the frontend to remove it from your zigbee2mqtt configuration. To be sure, also restart zigbee2mqtt after removing the bulb.

1. Make sure the bulb is acting on it's own behalf.
Do a  factory reset. This is to make sure the bulb is not known to any Hue bridge, app or zigbee group.

Now add the bulb back into zigbee2mqtt.

After the bulb was added, do the following:
1. Go to z2m frontend.
1. Go to the `Devices` tab and select your bulb.
1. Go to the `Exposes` tab.
1. Click all the 'refresh' buttons you can see, one by one, leave some time in between presses.
1. Share screenshot of the `Exposes` tab.
1. Share screenshot of the `State` tab.
1. Turn on the bulb.
1. Set color temperature to coolest.
2. Turn the bulb OFF.
1. Set color temperature to warmest.
2. Turn the bulb on.
3. Share what you see.

Grt,

**Comment by R3Cardoso** [comment]
I will not remove and add the light, because I just done a clear zigbee2mqtt installation last week, after changing to new adapter.

5.  ![Image](https://github.com/user-attachments/assets/cd5c07fe-9669-4402-8584-06cd1a52bb75)
6.  ![Image](https://github.com/user-attachments/assets/9cb92dc8-18f2-4aab-bef4-f5be53f3d29d)
7.  ![Image](https://github.com/user-attachments/assets/6f97d8ec-5323-4842-81c8-69f632fcff43)
     ![Image](https://github.com/user-attachments/assets/348c8bde-b013-4b6d-b6e2-76b1597ea51d)
9. and 10.  ![Image](https://github.com/user-attachments/assets/ab7ac149-c465-494d-9aad-ccedb68d5856)
                  ![Image](https://github.com/user-attachments/assets/2090dd9f-789c-4a80-b475-8499c6d04ca7)
11. ![Image](https://github.com/user-attachments/assets/b7328620-ee77-4e09-91f3-012bacd9df5f)
      ![Image](https://github.com/user-attachments/assets/257cf282-4401-46e4-8312-a88aab986305)
(important information, the light color stay coolest after turn on, instead of warmest as it should be.)

13. [45df7312_zigbee2mqtt_2025-03-19T22-14-47.821Z.log](https://github.com/user-attachments/files/19352942/45df7312_zigbee2mqtt_2025-03-19T22-14-47.821Z.log)

**Comment by djurny** [comment]
Hi @R3Cardoso,
Great! Thanks for the feedback.
Seems that `execute_if_off` is not yet reported as being active for your bulb.

Please do the following:
1. Go to zigbee2mqtt frontent.
1. Go to `Settings` (in the wheel dropdown).
1. Go to the `MQTT` tab.
1. Note down the `Base topic`.
1. Go to home assistant.
1. Open `Developer tools`, go to `Actions` tab.
1. Select `MQTT: Publish` in the action selector.
1. Fill out `zigbee2mqtt/name-of-your-bulb/set` in the `Topic`.
   The `zigbee2mqtt` in the `Topic` needs to be the `Base topic` from the zigbee2mqtt `Settings/MQTT` tab.
1. Check `Payload` and fill out:     
   ```
   {
     "color_options": {
       "execute_if_off": "false"
     }
   }
   ```
1. Uncheck `Retain`, `QoS` and `Evaluate payload`.
1. Click `Perform action`.
1. Confirm that zigbee2mqtt shows that something has been sent to the bulb.

Please repeat the previous actions, turn on light, set to warmest, turn off, set to coolest, turn on light. It should now turn on with coolest `color_temp`. 

Repeat in reverse to confirm; turn on light, set to coolest, set to warmest, set to coolest, turn off, set to warmest, turn on light. It should not turn on with warmest `color_temp`.

Grt,


---
For reference, how the above will look for one of my Hue bulbs:
![Image](https://github.com/user-attachments/assets/f4c030b3-3ea1-45e8-954b-dfaa864e7ce1)

Logging of zigbee2mqtt when I send the `execute_if_off` setting to one of my Hue bulbs:
![Image](https://github.com/user-attachments/assets/b287b345-7554-4a1d-8768-7e2006395229)
Focus on the `"color_options": { "execute_if_off": true }` part.

The `State` as reported by zigbee2mqtt frontend for one of my Hue bulbs after publishing the `execute_if_off` payload:
```
{
    "brightness": 254,
    "color_mode": "color_temp",
    "color_options": {
        "execute_if_off": "true"
    },
    "color_temp": 281,
    "color_temp_startup": 366,
    "last_seen": "2025-03-19T17:20:01-07:00",
    "level_config": {
        "execute_if_off": true
    },
    "linkquality": 65,
    "power_on_behavior": "on",
    "state": "ON",
    "update": {
        "installed_version": 16786690,
        "latest_version": 16786690,
        "state": "idle"
    },
    "effect": null
}
```

**Comment by R3Cardoso** [comment]
So, yes it works fine now when turning the light on in zibgee2mqtt, if I set coolest > turn off > set warmest > turn on, the light turns on already on warmest.
The state also show the execute_if_off": "true

But doing it using my automations and AL, it still adapts only after turning on.

Runing you automation code you share also doesn't help.

About my automation codes to turn on lights, I use virtual helpers to turn on lights using "action: adaptive_lighting.apply".

**Comment by djurny** [comment]
Hi @R3Cardoso,
> So, yes it works fine now 

Well now, that is good to hear.

Next would be to check why your automations and helpers (scripts?) seem to behave differently...

Grt,

**Comment by R3Cardoso** [comment]
I also set this setting on zigbee2mqtt:

> homeassistant:
>   legacy_entity_attributes: true
> mqtt:
>   force_disable_retain: false
> advanced:
>   cache_state: true
>   cache_state_persistent: true
>   cache_state_send_on_startup: false
>   report: true
>   legacy_api: false
> device_options:
>   disabled: false
>   homeassistant: true
>   optimistic: true
>   filtered_cache: ["brightness", "rgb_color", "hs_color", "xy_color", "color_temp_kelvin", "color_temp", "color", "color_mode"]

About my automation, here it is:

> triggers:
>   - entity_id:
>       - input_boolean.switchhelper
>     to: "on"
>     id: "ON"
>     trigger: state
>   - entity_id:
>       - input_boolean.switchhelper
>     to: "off"
>     id: "OFF"
>     trigger: state
> actions:
>   - choose:
>       - conditions:
>           - condition: trigger
>             id:
>               - "ON"
>         sequence:
>           - data:
>               turn_on_lights: true
>               adapt_brightness: true
>               adapt_color: true
>               entity_id: switch.adaptive_lighting_setting
>               lights:
>                 - light.ceiling
>             action: adaptive_lighting.apply
>       - conditions:
>           - condition: trigger
>             id:
>               - "OFF"
>         sequence:
>           - data:
>               entity_id: light.ceiling
>               transition: 1
>             action: light.turn_off
> mode: restart

AL settings transition are set to zero.

**Comment by djurny** [comment]
Hi there,
Just to let you know, I also use `adaptive_lighting.apply` to turn on all my AL-controlled `light` groups and all works well. So, you could for example test if `adaptive_lighting.apply` has the same behavior as the `State` to `ON` flip in the z2m frontend. 

> AL settings transition are set to zero.

That might explain why you see the lights turning on instantaneously.

For my configuration I could test it as below as I turn on any light with the following settings:
![Image](https://github.com/user-attachments/assets/d7be37c5-9acb-40b7-8ea2-9ddb5e7abde0)

Or as below to emulate the z2m frontend action:
![Image](https://github.com/user-attachments/assets/5795f83b-4566-4f69-a82f-fb56acb2b2f9)

Grt,

**Comment by R3Cardoso** [comment]
I doesn't work with either options. The light turns on last option and adapts during first 1 or 2 seconds.
don't know if the problem is related to zigbee2mqtt device, as when  I run the automation you made, zigbee2mqtt only looks for light first name and fires a error.
But I already change one like to have a continuos name and still doesn't work.

**Comment by djurny** [comment]
Hi @R3Cardoso,
Well, priming the `color_temp` for your bulbs using the `execute_if_off` option for sure works, as you have demonstrated to yourself earlier on. 

> zigbee2mqtt only looks for light first name and fires a error.

If I copy the [gist](https://gist.github.com/djurny/f3e8dfc429f48d312ac172f5ac5a2c01)  and paste it into an automation in hass, it works without error. Not sure how you can make it do what you say it does for you. I can only guess you have a typo somewhere?
Grt,

**Comment by JDFS404** [comment]
@djurny Do I understand correctly that your YAMLs deal with the 1 sec transition which occurs with Hue lights (I only have Hue lights in my home through Z2M), which is very annoying when let's say a light hasn't been on for a couple of hours and first it blinds you and then will adapt to the AL brightness / color temp?

**Comment by djurny** [comment]
Hi @JDFS404,
The YAMLs should take care of setting the `color_temp` of Zigbee controlled lights (which support the `execute_if_off` `color_option`) in normal use cases.
The YAMLs do not take care of any `brightness` setting.

May I ask how you turn your lights on/off?

Groetjes,

**Comment by JDFS404** [comment]
Ok duidelijk - hier even in het NL omdat wellicht wat dingen lost in translation gaan. Ik gebruik AppDaemon (ControllerX) icm Hue Wall Modules om licht (groepen) aan en uit te doen, maar kan het net zo goed via voice (Google Assistant) als via het HA dashboard doen.

Wat ik vaak merk, is dat als ik de lichten aan doe (al mijn lampen staan op AL) dat die op het oog aangaat op de laatste stand, en dan in 1 sec schiet naar de AL settings op dat moment. Soms kan ik dus flink wat kleurverschil en nog wat brightness verschil hebben. Of ligt dat dan aan mijn config? Zie hier bijv. een config:
```- name: Attic
  lights:
    - light.attic
  min_brightness: 30
  max_brightness: 90
  prefer_rgb_color: true
  detect_non_ha_changes: true
  min_color_temp: 2000
  max_color_temp: 6500
  interval: 30
```
EN:
Ok, clear – just writing this in Dutch because some things might get lost in translation. I’m using AppDaemon (ControllerX) in combination with Hue Wall Modules to turn lights (or groups of lights) on and off, but I can just as easily do it via voice (Google Assistant) or through the HA dashboard.

What I often notice is that when I turn on the lights (all my bulbs are set to Adaptive Lighting), they *appear* to turn on at the last state, and then within 1 second they jump to the current Adaptive Lighting settings. This sometimes causes noticeable color changes and some brightness difference. Or is that due to my config? For example, here's a config:

```yaml
- name: Attic
  lights:
    - light.attic
  min_brightness: 30
  max_brightness: 90
  prefer_rgb_color: true
  detect_non_ha_changes: true
  min_color_temp: 2000
  max_color_temp: 6500
  interval: 30
```

**Comment by djurny** [comment]
Hi @JDFS404,
Note that the YAMLs provided have not been tested with `prefer_rgb_color`, the setting of the `color_temp` is done with `color_temp` in Kelvin, not RGB. If it does work, there might be some gap or transition in case the Kelvin value to RGB mapping is not accurate.

I'm also not familiar with AppDaemon or ControllerX, so I cannot comment on those unfortunately. I can only assume the behavior you are seeing is similar to using bound switches to the bulbs, so that both HASS nor z2m are not aware of the lights turning on - and therefor cannot intercept anything - as soon as the light itself announces that it has been turned on.

In those cases, when switching on lights outside of HASS, adaptive lighting cannot do much until HASS realizes the lights have been turned on. I think AL gets notified by HASS on this event and then decides when to start adapting - this is the `adapt_delay` if I'm not mistaken. Most of the seamlessness of AL depends on it's ability to `intercept` calls to the lights by HASS, so it can already set the parameters you'd expect at `turn_on`.

For the `color_temp` the YAMLs can provide relief (in normal use cases), but for the brightness I'm not sure. There is some `execute_if_off` for `level_config` and `current_level_startup`, but I think that is for a 'cold' turn on when power is cycled, not when turning off/on the light using Zigbee.

Not sure if that answers your question,
Groetjes,

**Comment by R3Cardoso** [comment]
You don't need to set a autoamtion to keep send it, lights when one will adapt normally, this situation is for when you turn it on from be off during a long time.
Instead adapt after turn on, it adapts before.

edit: was not here I paste the solution I found for this situation, let me search for it.

**Comment by djurny** [comment]
> Hi [@djurny](https://github.com/djurny) I took a fresh break and re-read everything. With the `execute_if_off` enabled, I'm able to send the color temp to my Hue lights - I only have Hue lights so that's a plus.
> 
> I haven't yet set up your automations to do it every x minutes, because I probably have to repeat it for every light (all my AL instances manage a HA group of lights).
> 
> I hope that one day this will all be fixed in the back-end, but for now I'll do it like this :D

Hi @JDFS404 ,
Nice! I have not touched that `automation` thing for a while. It's useful to prepare ("prime") your lights before they turn on, without the need for separate scripts per light (or AL instance) you want to turn on. It will also make sure your lights will have the proper `color_temp` if you decide to turn them on via other means, like through the z2m frontend or the actual `light.turn_on` command.
Groetjes,

**Comment by R3Cardoso** [comment]
@JDFS404 do as I said, it works fine.
I also use lights and switch from philips hue.

I can share the automation I use for it

**Comment by Kuchiru** [comment]
Hi everyone, im happy to report that this has been resolved in Z2M, even for bulbs that do not support execute if off.
Enable the newly introduced Hue Native Control in the "Settings (specific)" tab on your hue bulbs and it'll just work.
This also fixes more than just this behavior, see the pull request below for additional info :)

Big props to [Mihonarium](https://github.com/Mihonarium)
Pull request(merged): https://github.com/Koenkk/zigbee-herdsman-converters/pull/11655

edit: groups dont work.. yet?

**Comment by JDFS404** [comment]
> Hi everyone, im happy to report that this has been resolved in Z2M, even for bulbs that do not support execute if off. Enable the newly introduced Hue Native Control in the "Settings (specific)" tab on your hue bulbs and it'll just work. This also fixes more than just this behavior, see the pull request below for additional info :)
> 
> Big props to [Mihonarium](https://github.com/Mihonarium) Pull request(merged): [Koenkk/zigbee-herdsman-converters#11655](https://github.com/Koenkk/zigbee-herdsman-converters/pull/11655)
> 
> edit: groups dont work.. yet?

Sweet! I saw your reply on Reddit, I'm about to try this out. Do you mind writing down the automations you're currently using for the `off` state?

**Comment by Kuchiru** [comment]
> > Hi everyone, im happy to report that this has been resolved in Z2M, even for bulbs that do not support execute if off. Enable the newly introduced Hue Native Control in the "Settings (specific)" tab on your hue bulbs and it'll just work. This also fixes more than just this behavior, see the pull request below for additional info :)
> > Big props to [Mihonarium](https://github.com/Mihonarium) Pull request(merged): [Koenkk/zigbee-herdsman-converters#11655](https://github.com/Koenkk/zigbee-herdsman-converters/pull/11655)
> > edit: groups dont work.. yet?
> 
> Sweet! I saw your reply on Reddit, I'm about to try this out. Do you mind writing down the automations you're currently using for the `off` state?

Hope your findings are as positive as mine!
Do you mean like the automations that prime the color temperature? these are no longer needed with this enabled.

The behavior of Z2M groups doesn't seem to be working quite right at this moment, i was able to get good behavior with a home assistant group instead. there is a pull request in draft for targeting group behavior that could fix this.

**Comment by JDFS404** [comment]
@Kuchiru can confirm indeed that for all my Hue lights it now works as intended, FINALLY! @djurny many and massive thanks for an education into Z2M and automation logic 💯 

I'm only using HA groups, tried Zigbee groups once but it felt finnicky sometimes with HA and the logic I'm using.

**Comment by stephanGarland** [comment]
May or may not be of relevance to anyone here, but I ran into the same problem, except since my bulbs were directly bound to their switches, AL couldn't intercept the turn_on calls quickly enough, so I would always get a visible color/brightness shift. I wound up making a cron automation that sends the current AL settings to bulbs that are off, so that when you turn them on, they're always correct, and [published it here](https://github.com/stephanGarland/ha-al-prestage).
